bakery/service_test.go

```diff
@@ -131,10 +143,93 @@ func (s *ServiceSuite) TestMacaroonPaperFig6FailsWithBindingOnTamperedSignature(
 	}
 
 	// client makes request to ts.
-	err = ts.Check(d, strCompFirstPartyChecker(""))
+	err = ts.Check(d, strcmpChecker(""))
 	c.Assert(err, gc.ErrorMatches, "verification failed: signature mismatch after caveat verification")
 }
 
+func (s *ServiceSuite) TestNeedDeclared(c *gc.C) {
+	locator := make(bakery.PublicKeyLocatorMap)
+	firstParty := newService(c, "first", locator)
+	thirdParty := newService(c, "third", locator)
+
+	// firstParty mints a macaroon with a third-party caveat addressed
+	// to thirdParty with a need-declared caveat.
+	m, err := firstParty.NewMacaroon("", nil, []checkers.Caveat{
+		checkers.NeedDeclaredCaveat(checkers.Caveat{
+			Location:  "third",
+			Condition: "something",
+		}, "foo", "bar"),
+	})
+	c.Assert(err, gc.IsNil)
+
+	// The client asks for a discharge macaroon for each third party caveat.
+	d, err := bakery.DischargeAll(m, func(_ string, cav macaroon.Caveat) (*macaroon.Macaroon, error) {
+		return thirdParty.Discharge(strcmpChecker("something"), cav.Id)
+	})
+	c.Assert(err, gc.IsNil)
+
+	// The required declared attributes should have been added
+	// to the discharge macaroons.
+	declared := checkers.InferDeclared(d)
+	c.Assert(declared, gc.DeepEquals, checkers.Declared{
+		"foo": "",
+		"bar": "",
+	})
+
+	// Make sure the macaroons actually check out correctly
+	// when provided with the declared checker.
+	err = firstParty.Check(d, checkers.New(declared))
+	c.Assert(err, gc.IsNil)
```

Comment:
I feel like maybe this should be `NotNil`. If you're going to use empty-string as a catch-all for a missing declaration, maybe it should not pass the `needs-declared` check.